### ISSUE TITLE
'vmlinux': hash table nchain (2053) differs from symbol count derived from SHT_DYNSYM section header

### ISSUE THREAD

**Issue by nickdesaulniers** [ClangBuiltLinux/linux#1157]
```
+ llvm-readelf --string-dump=.comment vmlinux
llvm-readelf: warning: 'vmlinux': hash table nchain (2053) differs from symbol count derived from SHT_DYNSYM section header (2436)
```
s390 kernels built from mainline produce the above.

**Comment by nickdesaulniers** [comment]
cc @uweigand

**Comment by uweigand** [comment]
This is very weird, given that a vmlinux image is not dynamically linked and therefore should have neither a DT_HASH hash table nor a SHT_DYNSYM dynamic symbol table.

What linker (and command line) was used to build this vmlinux image?   Can you post the output of "readelf -a vmlinux"?

**Comment by uweigand** [comment]
Hmm, I can reproduce the issue myself.   I'll have a look.

**Comment by uweigand** [comment]
Turns out the reason for the presence of .dynsym etc seems to be that the kernel was linked using -pie to enable KASLR.  So that part is actually OK.

However, the problem is on the side of the llvm-readelf tool, which assumes the size of each entry in .hash is 4 bytes, when it is actually 8 bytes on s390x (and alpha) for historical reasons.   I've opened a bug against llvm-readelf here:
https://llvm.org/pr47681
